internal/deployment_client.go

```diff
@@ -0,0 +1,244 @@
+	// safely if user does not query closed workflows.
+	// NOTE: Experimental
+	DeploymentReachabilityClosedWorkflows
+
+	// DeploymentReachabilityUnreachable - The deployment is not reachable by
+	// any workflow because all the workflows who needed this
+	// deployment are out of the retention period. The deployment can be
+	// decommissioned safely.
+	// NOTE: Experimental
+	DeploymentReachabilityUnreachable
+)
+
+type (
+	// Deployment identifies a set of workers. This identifier combines
+	// the deployment series name with their Build ID.
+	// NOTE: Experimental
+	Deployment struct {
+		// SeriesName - Name of the deployment series. Different versions of the same worker
+		// service/application are linked together by sharing a series name.
+		SeriesName string
+
+		// BuildID - identifies the worker's code and configuration version.
+		BuildID string
+	}
+
+	// DeploymentTaskQueueInfo describes properties of the Task Queues involved
+	// in a deployment.
+	// NOTE: Experimental
+	DeploymentTaskQueueInfo struct {
+		// Name - Task queue name.
+		Name string
+
+		// Type - The type of this task queue.
+		Type TaskQueueType
+
+		// FirstPollerTime - Time when the server saw the first poller for this task queue
+		// in this deployment.
+		FirstPollerTime time.Time
+	}
+
+	// DeploymentInfo holds information associated with
+	// workers in this deployment.
+	// Workers can poll multiple task queues in a single deployment,
+	// which are listed in this message.
+	// NOTE: Experimental
+	DeploymentInfo struct {
+		// Deployment - An identifier for this deployment.
+		Deployment Deployment
+
+		// CreateTime - When this deployment was created.
+		CreateTime time.Time
+
+		// IsCurrent - Whether this deployment is the current one for its deployment series.
+		IsCurrent bool
+
+		// TaskQueuesInfo - List of task queues polled by workers in this deployment.
+		TaskQueuesInfo []DeploymentTaskQueueInfo
+
+		// Metadata - A user-defined set of key-values. Can be updated with [DeploymentClient.SetCurrent].
+		Metadata map[string]*commonpb.Payload
+	}
+
+	// DeploymentListEntry is a subset of fields from DeploymentInfo
+	// NOTE: Experimental
+	DeploymentListEntry struct {
+		// An identifier for this deployment.
+		Deployment Deployment
+
+		// When this deployment was created.
+		CreateTime time.Time
+
+		// Whether this deployment is the current one for its deployment series.
+		IsCurrent bool
+	}
+
+	// DeploymentListIterator is an iterator for deployments.
+	// NOTE: Experimental
+	DeploymentListIterator interface {
+		// HasNext - Return whether this iterator has next value.
+		HasNext() bool
+
+		// Next - Returns the next deployment and error
+		Next() (*DeploymentListEntry, error)
+	}
+
+	// DeploymentListOptions are the parameters for configuring listing deployments.
+	// NOTE: Experimental
+	DeploymentListOptions struct {
+		// PageSize - How many results to fetch from the Server at a time.
+		// Optional: defaulted to 1000
+		PageSize int
```

Comment:
Is this needed?

Reply:
It is an optimization to reduce the number of roundtrips vs memory available. Frankly, doubt many people will use it, but it was useful at least for testing (forcing several roundtrips...).

Reply:
Hrmm, wonder if we should remove it. Will defer to others' opinions here.

Reply:
We have it in ScheduleListOptions, I think we should be consistent...

Reply:
We don't for `GetWorkflowHistory` which returns an iterator (even though we could), so we're going to be inconsistent regardless. But I am ok with leaving it.